### PR TITLE
Added proposal for Read-Write coordination-free bucket operation contract

### PR DESCRIPTION
**Pull request by bwplotka**
Main goals are:
* Support for eventual consistent object storages: https://github.com/improbable-eng/thanos/issues/298, https://github.com/improbable-eng/thanos/issues/377
* Avoid race condidtions on compaction and retention: https://github.com/improbable-eng/thanos/issues/564, https://github.com/improbable-eng/thanos/issues/271
* Handle partial uploads gracefully: https://github.com/improbable-eng/thanos/issues/377

CC @improbable-ludwik

### CODE REVIEW

File: docs/proposals/read-write-operations-bucket.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+* Thanos compactor uploads compacted block and deletes source blocks. After next sync iteration it does not see a new block (read after write eventual consistency). It sees gap, wrongly plans next compaction and causes non-resolvable overlap.
+* Thanos compactor uploads compacted block and deletes source blocks. Thanos Store Gateway syncs every 3m so it missed that fact. Next query that hits store gateway tries to fetch deleted source blocks and fails.
+
+Currently, we have only basic safeguards against those in form of [`syncDelay`](https://github.com/improbable-eng/thanos/blob/bc088285a1b4bf464fdf2539e4b365b805874eed/pkg/compact/compact.go#L187). 
+That helps with operations between sidecar and compactor but does not fix inter-compactor logic in case of eventual consistent storage or partial upload failures.
+
+We also partially assumed strong consistency for object storage, because [GCS has strong consistency](https://cloud.google.com/storage/docs/consistency) and S3 has as it as well,
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: `and S3 has as it as well`  ->  `and S3 has it as well`

File: docs/proposals/read-write-operations-bucket.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+* Thanos compactor uploads compacted block and deletes source blocks. After next sync iteration it does not see a new block (read after write eventual consistency). It sees gap, wrongly plans next compaction and causes non-resolvable overlap.
+* Thanos compactor uploads compacted block and deletes source blocks. Thanos Store Gateway syncs every 3m so it missed that fact. Next query that hits store gateway tries to fetch deleted source blocks and fails.
+
+Currently, we have only basic safeguards against those in form of [`syncDelay`](https://github.com/improbable-eng/thanos/blob/bc088285a1b4bf464fdf2539e4b365b805874eed/pkg/compact/compact.go#L187). 
+That helps with operations between sidecar and compactor but does not fix inter-compactor logic in case of eventual consistent storage or partial upload failures.
+
+We also partially assumed strong consistency for object storage, because [GCS has strong consistency](https://cloud.google.com/storage/docs/consistency) and S3 has as it as well, 
+but with [some caveats](https://codeburst.io/quick-explanation-of-the-s3-consistency-model-6c9f325e3f82) that we met.
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: So, does it have strong consistency or not? Or more importantly, do you want to say that effort has been made to such that our use of S3 is within it's strongly consistent guarantees?

File: docs/proposals/read-write-operations-bucket.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+guarantee this assumption needs to go away.
+
+We have an explicit API for Series fetching in form of gRPC StoreAPI, but we lack the verbose API for other side: the object storage format with assumptions on behavior that will
+cover partial blocks, eventual consistency and syncing requirements, which is a goal for this proposal.
+
+## Proposal
+
+We propose defining those set of rules to solve A, B and C goals that will define the object storage format API (and how to operate with those objects):
+
+1. Block once uploaded is never edited (no file within ULID directory is removed, edited or added).
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: `A block, once`

File: docs/proposals/read-write-operations-bucket.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+1. Block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+   to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+   still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+1. TSDB `meta.json` file within block *have to be* uploaded as first one.
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: Since we don't have strongly consistent, why  does it matter that it is the first to be uploaded?

File: docs/proposals/read-write-operations-bucket.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+* Block removal 
+
+As those operations are not atomic itself, partial blocks or inconsistent block view may occur for short time (read-after-write consistency).
+To mitigate this we introduce second rule:
+
+ > 2 . Block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+
+Readers may attempt to load block's files, but should be aware that those files can be partially uploaded (e.g meta.json to grab external labels).
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: Is it that the files are partially uploaded, or that the set of the files uploaded will not be necessarily complete initially? I understand it's the latter, and if so `meta.json` can hold data that is a checksum of the set of files(names), which in turn could let the reader know early if the data is fully available.

File: docs/proposals/read-write-operations-bucket.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+*Newer block from two or more overlapping blocks fully submatches the source blocks of older blocks. Older blocks can be then ignored.*
+
+Fully word is crucial. For example we won't be able to resolve case with block ABCD and CDEF. This is because there is no logic for decompact or vertical compaction.
+
+Having this kind of overlap support, we can delay deletion by forming 6th rule:
+
+> 6 . Adding "replacement" or "overlapped" block causes old blocks to be scheduled for removal when new block is ready by compactor (`HealthyOverlapped` state). [Compactor/Repair change needed]
+
+This rule ensures that we can detect when to delete block, and we delete it only once new block is ready, so when we mitigated eventual consistency: all readers should be aware of new block.
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: `so when we mitigated eventual consistency` , do you mean , eg: `this will mitigate eventual consistency`?

File: docs/proposals/read-write-operations-bucket.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+* What if someone will set too low `syncDelay`? And upload itself will take more than `syncDelay` time. Block will be assumed as `ToDelete` state and will be removed. Thanks of `deleteDelay` it might have still some time to recover.How to ensure that will not happen? 
+  * Minimum syncDelay? 
+  * Grab time for syncDelay from object meta instead? Not from ULID? That is quite hard.
+  * Add safeguard on upload path? `upload timeout` has to smaller than `syncDelay`.
+
+* What if one block is malformed. Readers cannot handle it and crashes. How repair procedure will work? We can have repair process that can download block locally, rewrite it and fix it and upload. Problem is that it will take `syncDelay` time to
+appear in system. Since we are blocked, we need to make the block available immediately, ignoring eventual consistency limitations.
+Potential solutions:
+  * Just delete problematic block and live with 15 minute of delay for not having some portion of metrics? Will it not cause any downsampling issues?
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: If this would cause alerts to go off that should not, I would refrain from just pulling the data...

File: docs/proposals/read-write-operations-bucket.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+## Alternatives:
+
+### Require strong consistent object storage.
+
+As mentioned in #Motivation section this would block community adoption as only a few of object storages has strong or even
+ clearly stated consistency guarantees.
+
+### Full immutability 
+
+By not removing any object, we may simplify problem. However we need hard deletions are:
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: `we need hard deletions are:` -> `we need hard deletions because they are:`

File: docs/proposals/read-write-operations-bucket.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+- enormously reducing number of blocks in storage, improving syncing time.
+
+### Special integrity JSON file
+
+We may want to add some file to commit and mark integrity of the block. That file may hold some information like
+files we expect to have in the directory and maybe some hashes of those. 
+
+This was rejected for we can reuse existing meta.json as "integrity" marker (e.g if after 15m we don't have meta file, it is partially uploaded and should be removed).
+
+Additional information like expected number of files can be added later if needed, however with rule no 3. we should not need that.
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: I think rule3 (meta file as first to be upload) may have to be refactored to talk about som content of that file that constitutes a checksum.

File: docs/proposals/read-write-operations-bucket.md (hunk 2 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Yes, indeed. There was an effort to eliminate invokes `Get` or `Exists` before writing object in crucial parts, but we DID not eliminate them everywhere, that's why some `syncDelay` exists for only part of new blocks.

File: docs/proposals/read-write-operations-bucket.md (hunk 4 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Very good point, I will explain it further with:

>  It matters even when we don't have strong consistency, because 
it helps us detect the case of `upload` crashed in between of upload. We only upload meta.json after successful uploads of rest of the files for the block.

File: docs/proposals/read-write-operations-bucket.md (hunk 7 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: What do you mean? (:

File: docs/proposals/read-write-operations-bucket.md (hunk 9 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Responded above, I think checksum is not helpful here

File: docs/proposals/read-write-operations-bucket.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+ which will affect performance.
+ * Does not help with partial upload and syncing issues
+ 
+ As vertical compaction might be something we can support in future, it clearly does not help with problems we stated here.
+ 
+### Do not introduce `deleteDelay`
+ 
+We may avoid introducing additional state, by adding mitigation for not having delayed removal.
+
+For example for retention apply or manual block deletion, when we would delete block immediately we can have query failures (object does not exists for getrange operations)
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: Is retention apply a know operation? This doesn't quite read right...

File: docs/proposals/read-write-operations-bucket.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+ 
+We can avoid it by:
+* Adding `retention` flag to readers (compactor have it, but store gateway does not). Cons: 
+  * This will increase configuration burden as you need to configure  retention for each resolution on each reader.
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: ` configure  retention` two spaces...

File: docs/proposals/read-write-operations-bucket.md [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+ 
+We may avoid introducing additional state, by adding mitigation for not having delayed removal.
+
+For example for retention apply or manual block deletion, when we would delete block immediately we can have query failures (object does not exists for getrange operations)
+ 
+We can avoid it by:
+* Adding `retention` flag to readers (compactor have it, but store gateway does not). Cons: 
+  * This will increase configuration burden as you need to configure  retention for each resolution on each reader. 
+  * This does not cover other deletion reasons, like manual removal (e.g for repair)
+* Handle not exists errors for queries against store gateway gracefully. For instance, if we have block loaded and bucket operation will return object not exists. We return no data.Cons:
```
Comment (improbable-ludwik) [inline docs/proposals/read-write-operations-bucket.md]: `data.Cons:` -> `data. Cons:` ?

File: docs/proposals/read-write-operations-bucket.md (hunk 10 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Yes, and actually the method is called like this. We can reword this.. How would you call it differently?

**Comment by bwplotka** [comment]
BTW, new TSDB code includes Parents as well!

```
func (c *LeveledCompactor) Write(dest string, b BlockReader, mint, maxt int64, parent *BlockMeta) (ulid.ULID, error) {
	entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
	uid := ulid.MustNew(ulid.Now(), entropy)

	meta := &BlockMeta{
		ULID:    uid,
		MinTime: mint,
		MaxTime: maxt,
	}
	meta.Compaction.Level = 1
	meta.Compaction.Sources = []ulid.ULID{uid}

	if parent != nil {
		meta.Compaction.Parents = []BlockDesc{
			{ULID: parent.ULID, MinTime: parent.MinTime, MaxTime: parent.MaxTime},
		}
	}

	err := c.write(dest, meta, b)
	if err != nil {
		return uid, err
	}

	level.Info(c.logger).Log("msg", "write block", "mint", meta.MinTime, "maxt", meta.MaxTime, "ulid", meta.ULID)
	return uid, nil
}

```

File: docs/proposals/read-write-operations-bucket.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+> 3 . TSDB `meta.json` file within block *have to be* uploaded as first one.
+
+Currently and as proposed `meta.json` is our integrity marker (it is uploaded as the last one). It matters even when we don't have strong consistency, because 
+it helps us detect the case of `upload` crashed in between of upload. We only upload meta.json after successful uploads of rest of the files for the block.
+
+With those rule one could say "why to wait 15 minutes if we can just check if meta.json is parsable". In eventual consistency world
+we cannot just check meta.json as even when it was uploaded at the end, other files might be still unavailable for a reader.
+
+Thanks of meta.json being uploaded at the end we can form 4th rule:
```
Comment (sbueringer) [inline docs/proposals/read-write-operations-bucket.md]: Maybe I just don't understand it, but in:
* in l.93 you stated meta.json is uploaded first
* in l.101 you stated meta.json being uploaded at the end

Comment (domgreen) [reply docs/proposals/read-write-operations-bucket.md]: Should indeed be the last thing was a typo earlier in the doc, think it will be corrected soon :+1:

File: docs/proposals/read-write-operations-bucket.md [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+## Proposal
+
+We propose defining those set of rules to solve A, B and C goals that will define the object storage format API (and how to operate with those objects):
+
+1. A block, once uploaded is never edited (no file within ULID directory is removed, edited or added).
+1. A block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+   to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+   still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+1. TSDB `meta.json` file within block *have to be* uploaded as first one.
+1. After 15 minutes (or time configured by `syncDelay`), if there is no fully loadable `meta.json`, we assume block is malformed, partially uploaded or marked for deletion and can be
```
[the remark this answers is not in the document]
Comment (domgreen) [reply docs/proposals/read-write-operations-bucket.md]: We are only deleting this from blob storage, the sidecar / ruler will then notice it has a local block that has not been successfully uploaded and re upload. Therefore the deleting from block storage allows us to re upload a block.

[the remark this answers is not in the document]
Comment (FUSAKLA) [reply docs/proposals/read-write-operations-bucket.md]: But the sidecar/ruler can meanwhile delete this block and than you have nothing to re upload or this cannot happen?

[the remark this answers is not in the document]
Comment (domgreen) [reply docs/proposals/read-write-operations-bucket.md]: I think this would only happen if you have a really aggressive clean up policy in your Prometheus configuration ... based on the default sync / delete timings it would attempt to re upload in about 30mins.

File: docs/proposals/read-write-operations-bucket.md (hunk 13 above)
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Yea, it is a typo sorry ;p Meta.json should be uploaded as last

File: docs/proposals/read-write-operations-bucket.md [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+   *removed* by compactor. [Compactor change needed]
+1. Overlaps (in external labels and time dimensions) with the same data are allowed and should be handled by readers gracefully. [Compactor change needed]
+1. Adding "replacement" or "overlapped" block causes old blocks to be scheduled for removal when new block is ready by compactor (`HealthyOverlapped` state). [Compactor/Repair change needed]
+1. To schedule delete operation, delete `meta.json` file. All components will exclude this block and compactor will do eventual deletion assuming the block is partially uploaded. [Compactor change needed]
+1. Compactor waits minimum 15m (`deleteDelay`) before deleting the whole `To Delete` block. [Compactor change needed]
+
+Below diagram shows the block states defined by our rules.
+
+[![Diagram](https://docs.google.com/drawings/d/e/2PACX-1vSZjddrWBkf2ynczeIMSzGo-yqe7Cg9X-TG_LjyWAP_ToyZolMEG_UnbNlvoSetRhDLmlfABd_rJ4W9/pub?w=960&h=720)](https://docs.google.com/drawings/d/1leRN6GUvU98Ur2TowMefk4ycTFNBRi9t1-dGtOPmHg0/edit?usp=sharing)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: yes, that's correct

[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: fixed diagram

File: docs/proposals/read-write-operations-bucket.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+
+Having this kind of overlap support, we can delay deletion by forming 6th rule:
+
+> 6 . Adding "replacement" or "overlapped" block causes old blocks to be scheduled for removal when new block is ready by compactor (`HealthyOverlapped` state). [Compactor/Repair change needed]
+
+This rule ensures that we can detect *when* to delete block, and we delete it only once new block is ready, so when all readers already loaded new block (e.g to syncDelay).
+
+This also means that repairing block gracefully (rewriting for deleting series or other reason that does not need to be unblocked), is as easy as uploading new block. Compactor will delete older overlapped block eventually.
+
+There is caveat for rule 2th rule (block being ready) for compactor. In case of compaction process we compact and we want to be aware of this block later on. Because of
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: omg :man_facepalming:

File: docs/proposals/read-write-operations-bucket.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+  * Use `ignore_delay` option and avoid Fresh state as possible. Eventual consistency issue may hit us.
+ 
+* Do we allow repair procedure when compactor is running? This is quite unsafe as compaction/downsamlping operation might be in progress 
+so repair procedure might be avoided. Turning off compactor might be tricky as well in potential HA mode. Maybe we want to introduce 
+"locking" some external labels from compactor operations. If yes, how to do it in eventual consistency system?
+
+## Action Items:
+
+*Ensure our block operations are following those rules (e.g create methods that ensures this).
+*Tests edge cases.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: good point - moved to TBD as we would like to define more specific action plan IMO

File: docs/proposals/read-write-operations-bucket.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/read-write-operations-bucket.md]: Well we are serious on this point as well. But without meta.json this block is kind of useless, as it does not have external label.

This case happens a lot for compactor. If it is uploading and crashes in the middle it ends up with some weird partial block.  Since operation of upload failed, nothing got removed, and compactor will just retry. So it is completely safe to remove such partial block.

The other writer is sidecar/ruler. In this case we need to ensure sidecar and rule will retry as well in case of upload being failed. This is done by writing to persistent file once upload is succesful. If not `shipper` package will retry until success.

**Comment by bwplotka** [comment]
Addressed comments, Action Items TBD.

I feel like `deleteSync` is crucial here. See first risk in #Risk section.

File: docs/proposals/read-write-operations-bucket.md [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+appear in system. Since we are blocked, we need to make the block available immediately, ignoring eventual consistency limitations.
+Potential solutions:
+  * Just delete problematic block and live with 15 minute of delay for not having some portion of metrics? Will it not cause any downsampling issues?
+  * Use `ignore_delay` option and avoid Fresh state as possible. Eventual consistency issue may hit us.
+ 
+* Do we allow repair procedure when compactor is running? This is quite unsafe as compaction/downsamlping operation might be in progress 
+so repair procedure might be avoided. Turning off compactor might be tricky as well in potential HA mode. Maybe we want to introduce 
+"locking" some external labels from compactor operations. If yes, how to do it in eventual consistency system?
+
+*
```
Comment (domgreen) [inline docs/proposals/read-write-operations-bucket.md]: remove line
